zebra-state/src/service/write.rs

```diff
@@ -108,31 +176,105 @@ pub fn write_blocks_from_channels(
+
+                // If the block is invalid, reject any descendant blocks.
+                //
+                // At this point, we know that the block and all its descendants
+                // are invalid, because we checked all the consensus rules before
+                // committing the block to the non-finalized state.
+                // (These checks also bind the transaction data to the block
+                // header, using the transaction merkle tree and authorizing data
+                // commitment.)
+                if let Some(parent_error) = parent_error {
+                    tracing::trace!(
+                        ?child_hash,
+                        ?parent_error,
+                        "rejecting queued child due to parent error"
+                    );
+                    result = Err(parent_error.clone());
+                } else {
+                    tracing::trace!(?child_hash, "validating queued child");
+                    result = validate_and_commit_non_finalized(
+                        &finalized_state,
+                        &mut non_finalized_state,
+                        network,
+                        child,
+                    )
+                    .map_err(CloneError::from);
+
+                    if result.is_ok() {
+                        // Update the metrics if semantic and contextual validation passes
+                        metrics::counter!("state.full_verifier.committed.block.count", 1);
+                        metrics::counter!("zcash.chain.verified.block.total", 1);
+                    }
+                }
+
+                let _ = rsp_tx.send(result.clone().map(|()| child_hash).map_err(BoxError::from));
+
+                if let Err(ref error) = result {
+                    parent_error_map.insert(parent_hash, error.clone());
+                }
+            }
+
+            NonFinalizedWriteCmd::FinishProcessingQueued => {
+                // Clear out any errors from processed queued blocks without children
+                parent_error_map.clear();
 
-    // We're finished receiving non-finalized blocks from the state.
-    //
-    // TODO:
-    // - make the task an object, and do this in the drop impl?
-    // - does the drop order matter here?
-    non_finalized_block_write_receiver.close();
-    std::mem::drop(non_finalized_block_write_receiver);
+                while non_finalized_state.best_chain_len()
+                    > crate::constants::MAX_BLOCK_REORG_HEIGHT
+                {
+                    tracing::trace!("finalizing block past the reorg limit");
+                    let finalized_with_trees = non_finalized_state.finalize();
+                    finalized_state
+                        .commit_finalized_direct(finalized_with_trees, "best non-finalized chain root")
+                        .expect(
+                            "expected that errors would not occur when writing to disk or updating note commitment and history trees",
```

Comment:
This failure message is outdated, we should have updated it as part of the NU5 work:
```suggestion
                            "unexpected finalized block commit error: note commitment and history trees were already checked by the non-finalized state",
```